src/ostree/ot-builtin-archive-import.c

```diff
@@ -0,0 +1,259 @@
+                   GError **error)
+{
+  gboolean ret = FALSE;
+  const char *dot;
+  
+  dot = strrchr (s, '.');
+  if (!dot)
+    {
+      g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_INVALID_FILENAME,
+                           "Missing '.' in object filename");
+      goto out;
+    }
+
+  if (strcmp (dot, ".file") == 0)
+    *out_objtype = OSTREE_OBJECT_TYPE_FILE;
+  else if (strcmp (dot, ".dirtree") == 0)
+    *out_objtype = OSTREE_OBJECT_TYPE_DIR_TREE;
+  else if (strcmp (dot, ".dirmeta") == 0)
+    *out_objtype = OSTREE_OBJECT_TYPE_DIR_META;
+  else if (strcmp (dot, ".commit") == 0)
+    *out_objtype = OSTREE_OBJECT_TYPE_COMMIT;
+  else
+    {
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_INVALID_FILENAME,
+                   "Invalid object suffix: %s", dot);
+      goto out;
+    }
+
+  *out_checksum = g_strndup (s, dot - s);
+
+  ret = TRUE;
+ out:
+  return ret;
+}
+
+
+static gboolean
+import_object_from_zipfile (ArchiveState          *self,
+                            struct archive        *zipfile,
+                            struct archive_entry  *entry,
+                            GCancellable          *cancellable,
+                            GError               **error)
+{
+  gboolean ret = FALSE;
+  g_autoptr (GInputStream) instream = NULL;
+  const char *pathname = NULL;
+  g_autofree char *checksum = NULL;
+  OstreeObjectType objtype;
+
+  pathname = archive_entry_pathname (entry);
+  
+  if (!parse_object_name (pathname, &checksum, &objtype, error))
+    goto out;
+
+  if (OSTREE_OBJECT_TYPE_IS_META (objtype))
+    {
+      GByteArray *buf = g_byte_array_new ();
+
+      while (TRUE)
+        {
+          const guint8 *tmp_buf = NULL;
+          size_t read_size;
+          gint64 offset;
+          int r;
+          
+          r = archive_read_data_block (zipfile, (const void**)&tmp_buf, &read_size, &offset);
+          if (r == ARCHIVE_EOF)
+            break;
+          if (r < ARCHIVE_OK)
+            {
+              propagate_libarchive_error (error, zipfile);
+              goto out;
+            }
+
+          if (read_size == 0)
+            break;
+
+          g_byte_array_append (buf, tmp_buf, read_size);
+        }
+      
+      {
+        g_autoptr(GVariant) metadata
+          = g_variant_new_from_data (ostree_metadata_variant_type (objtype),
+                                     buf->data, buf->len, TRUE, NULL, NULL);
+        if (!ostree_repo_write_metadata_trusted (self->repo, objtype, checksum, metadata,
+                                                 cancellable, error))
+          goto out;
+      }
+
+      if (objtype == OSTREE_OBJECT_TYPE_COMMIT && self->commit_checksum == NULL)
+        {
+          self->commit_checksum = g_strdup (checksum);
+        }
+
+      g_byte_array_unref (buf);
+    }
+  else
+    {
+      g_autoptr(GInputStream) archive_stream = _ostree_libarchive_input_stream_new (zipfile);
+
+      if (!ostree_repo_write_content_trusted (self->repo,
```

Comment:
We shouldn't just trust the imput here, should we. It may lie about its checksum...